notify-send.sh

```diff
@@ -200,109 +200,172 @@ process_special_action() {
     action_key="$1"
     command="$2"
 
-    if [[ -z "$action_key" ]] || [[ -z "$command" ]]; then
+    if [ -z "$action_key" ] || [ -z "$command" ]; then
         echo "Command must not be empty"
         exit 1
     fi
 
     ACTION_COMMANDS=("${ACTION_COMMANDS[@]}" "$action_key" "$command")
 
-    if [[ "$action_key" != close ]]; then
+    if [ "$action_key" != close ]; then
         local action="$(make_action "$action_key" "$name")"
         ACTIONS=("${ACTIONS[@]}" "$action")
     fi
 }
 
 process_posargs() {
-    if [[ "$1" = -* ]] && ! [[ "$positional" = yes ]] ; then
-        echo "Unknown option $1"
-        exit 1
+    case "$1" in
+        -*) if [ "$positional" != yes ] ; then
+                echo "Unknown option $1"
+                exit 1
+            fi ;;
+    esac
+
+    if [ "$SUMMARY_SET" = n ]; then
+        SUMMARY="$1"
+        SUMMARY_SET=y
     else
-        if [[ "$SUMMARY_SET" = n ]]; then
-            SUMMARY="$1"
-            SUMMARY_SET=y
-        else
-            BODY="$1"
-        fi
+        BODY="$1"
     fi
 }
 
-while (( $# > 0 )) ; do
+while [ $# -gt 0 ] ; do
     case "$1" in
         -\?|--help)
             help
             exit 0
             ;;
+
         -v|--version)
             echo "${0##*/} $VERSION"
             exit 0
             ;;
-        -u|--urgency|--urgency=*)
-            [[ "$1" = --urgency=* ]] && urgency="${1#*=}" || { shift; urgency="$1"; }
-            process_urgency "$urgency"
+
+        -u|--urgency)
+            shift
+            process_urgency "$1"
+            ;;
+        --urgency=*)
+            process_urgency "${1#*=}"
             ;;
```

Comment:
These are a bit trickier to handle and, for some situations, the solution is clean.  However, in cases where the handling is a bit more convoluted than just checking the `--option=*` variant, one would probably need at least one extra repeated `case` statement and it starts polluting the option parsing section.

My recommendation would be replacing that kind of handling and validation with functions, much like this `process_urgency`, and sweep away all of that mess away from here.

Reply:
This is a bit of a more convoluted approach, but it works for any kind of `--option=*`-like option

```sh
while [ $# -gt 0 ]; do
  case "$1" in
    --[[:lower:]-][[:lower:]-]*=*)
      __opt="${1%=*}"
      __val="${1#*=}"
      shift
      set -- ignored "$__opt" "$__val" "$@"
      unset __opt __val
      ;;
  # other options
  esac
shift
done
```

Let's break it up:

1. The pattern `--[[:lower:]-][[:lower:]-]*=*` is equivalent to `--[-a-z]+=.*`, though some character classes take `LC_CTYPE` into consideration.  `[:lower:]` could be replaced with `a-z`;
2. We break the option and the value into temporary variables (`__opt` and `__val` resp.) for later usage;
3. `shift` the now parsed option
4. Use `set --` to set positional argument values:
   - We set the first one (`$1`) to some mock value;
   - `$2` and `$3` become the option and the value, respectively;
   - The rest is set as whatever still needs to be processed;
5. We clean up (`unset`) the temporary variables

What happens next is we break out of the case, the mock value we set gets shifted, and we are going to hit the case (if one exists) that covers option we parsed.  The rest is business as usual.

Say the following arguments are passed in

```sh
notify-send.sh --option=value summary body
```

Here, `$1` is `--option=value`, and `$@` would be `summary body`.  Once we parse `$1`, we get `__opt=--option` and `__val=value`.  We `shift`, resulting in the following situation:

```sh
notify-send.sh summary body
```

However, we then re`set` the positional arguments, putting them in the following state:

```sh
notify-send.sh ignored --option value summary body
```

Then, after we clean up and break out, the `ignored` mock argument is shifted and we end up with:

```sh
notify-send.sh --option value summary body
```

:tada: